File: indusmart-frontend/src/services/wishlistService.js
import axiosConfig from "../api/axiosConfig";

/**
 * Add Product To Wishlist
 */
const addToWishlist = async (buyerId, productId) => {

    const response = await axiosConfig.post(

        "/api/wishlist/add",

        null,

        {

            params: {

                buyerId,

                productId

            }

        }

    );

    return response.data;

};

/**
 * Remove Product From Wishlist
 */
const removeFromWishlist = async (

    buyerId,

    productId

) => {

    const response = await axiosConfig.delete(

        "/api/wishlist/remove",

        {

            params: {

                buyerId,

                productId

            }

        }

    );

    return response.data;

};

/**
 * Get Buyer's Wishlist
 */
const getWishlist = async (buyerId) => {

    const response = await axiosConfig.get(

        `/api/wishlist/${buyerId}`

    );

    return response.data;

};

/**
 * Check Product Exists In Wishlist
 */
const isWishlisted = async (

    buyerId,

    productId

) => {

    const response = await axiosConfig.get(

        "/api/wishlist/check",

        {

            params: {

                buyerId,

                productId

            }

        }

    );

    return response.data;

};

/**
 * Wishlist Count
 */
const wishlistCount = async (buyerId) => {

    const response = await axiosConfig.get(

        `/api/wishlist/count/${buyerId}`

    );

    return response.data;

};

/**
 * Clear Wishlist
 */
const clearWishlist = async (buyerId) => {

    const response = await axiosConfig.delete(

        `/api/wishlist/clear/${buyerId}`

    );

    return response.data;

};

/**
 * Product Wishlist Count
 */
const productWishlistCount = async (

    productId

) => {

    const response = await axiosConfig.get(

        `/api/wishlist/product-count/${productId}`

    );

    return response.data;

};

const wishlistService = {

    addToWishlist,

    removeFromWishlist,

    getWishlist,

    isWishlisted,

    wishlistCount,

    clearWishlist,

    productWishlistCount

};

export default wishlistService;

File: indusmart-backend/src/main/java/com/indusmart/service/WishlistService.java
package com.indusmart.service;

import com.indusmart.entity.Product;
import com.indusmart.entity.User;
import com.indusmart.entity.Wishlist;
import com.indusmart.repository.ProductRepository;
import com.indusmart.repository.UserRepository;
import com.indusmart.repository.WishlistRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class WishlistService {

    private final WishlistRepository wishlistRepository;
    private final UserRepository userRepository;
    private final ProductRepository productRepository;

    public WishlistService(WishlistRepository wishlistRepository,
                           UserRepository userRepository,
                           ProductRepository productRepository) {

        this.wishlistRepository = wishlistRepository;
        this.userRepository = userRepository;
        this.productRepository = productRepository;
    }

    /**
     * Add Product to Wishlist
     */
    public String addToWishlist(Long buyerId, Long productId) {

        Optional<User> buyerOptional =
                userRepository.findById(buyerId);

        if (buyerOptional.isEmpty()) {
            return "Buyer not found";
        }

        Optional<Product> productOptional =
                productRepository.findById(productId);

        if (productOptional.isEmpty()) {
            return "Product not found";
        }

        if (wishlistRepository.existsByBuyerIdAndProductId(
                buyerId,
                productId)) {

            return "Product already exists in wishlist";
        }

        Wishlist wishlist = Wishlist.builder()
                .buyer(buyerOptional.get())
                .product(productOptional.get())
                .build();

        wishlistRepository.save(wishlist);

        return "Product added to wishlist successfully";
    }

    /**
     * Remove Product from Wishlist
     */
    public String removeFromWishlist(Long buyerId,
                                     Long productId) {

        if (!wishlistRepository.existsByBuyerIdAndProductId(
                buyerId,
                productId)) {

            return "Product not found in wishlist";
        }

        wishlistRepository.deleteByBuyerIdAndProductId(
                buyerId,
                productId
        );

        return "Product removed from wishlist successfully";
    }

    /**
     * Get Buyer's Wishlist
     */
    public List<Wishlist> getWishlist(Long buyerId) {

        return wishlistRepository
                .findByBuyerIdOrderByAddedAtDesc(buyerId);
    }

    /**
     * Check Wishlist
     */
    public boolean isWishlisted(Long buyerId,
                                Long productId) {

        return wishlistRepository
                .existsByBuyerIdAndProductId(
                        buyerId,
                        productId
                );
    }

    /**
     * Wishlist Count
     */
    public long wishlistCount(Long buyerId) {

        return wishlistRepository.countByBuyerId(buyerId);
    }

    /**
     * Clear Wishlist
     */
    public String clearWishlist(Long buyerId) {

        wishlistRepository.deleteByBuyerId(buyerId);

        return "Wishlist cleared successfully";
    }

    /**
     * Product Wishlist Count
     */
    public long productWishlistCount(Long productId) {

        return wishlistRepository.countByProductId(productId);
    }

}